pkg/reconciler/taskrun/taskrun_expired.go

```diff
@@ -0,0 +1,181 @@
+func (tc *Reconciler) processTaskRunExpired(namespace, name string, tr *apispipeline.TaskRun) error {
+	klog.V(4).Infof("Checking if TaskRun %s/%s is ready for cleanup", namespace, name)
+	if tr.HasPipelineRunOwnerReference() {
+		return nil
+	}
+
+	if expired, err := tc.processTrTTL(tr); err != nil {
+		return err
+	} else if !expired {
+		return nil
+	}
+
+	// The TaskRun's TTL is assumed to have expired, but the TaskRun TTL might be stale.
+	// Before deleting the TaskRun, do a final sanity check.
+	// If TTL is modified before we do this check, we cannot be sure if the TTL truly expires.
+	// The latest TaskRun may have a different UID, but it's fine because the checks will be run again.
+	fresh, err := tc.PipelineClientSet.TektonV1alpha1().TaskRuns(namespace).Get(name, metav1.GetOptions{})
+	if errors.IsNotFound(err) {
+		return nil
+	}
+	if err != nil {
+		return err
+	}
+
+	if fresh.HasPipelineRunOwnerReference() {
+		return nil
+	}
+	// Use the latest TaskRun TTL to see if the TTL truly expires.
+	if expired, err := tc.processTrTTL(fresh); err != nil {
+		return err
+	} else if !expired {
+		return nil
+	}
+	// Cascade deletes the TaskRuns if TTL truly expires.
+	policy := metav1.DeletePropagationForeground
+	options := &metav1.DeleteOptions{
+		PropagationPolicy: &policy,
+		Preconditions:     &metav1.Preconditions{UID: &fresh.UID},
+	}
+	klog.V(4).Infof("Cleaning up TaskRun %s/%s", namespace, name)
+	return tc.PipelineClientSet.TektonV1alpha1().TaskRuns(fresh.Namespace).Delete(fresh.Name, options)
+}
+
+// processTTL checks whether a given TaskRun's TTL has expired, and add it to the queue after the TTL is expected to expire
+// if the TTL will expire later.
+func (tc *Reconciler) processTrTTL(tr *apispipeline.TaskRun) (expired bool, err error) {
+	// We don't care about the TaskRuns that are going to be deleted, or the ones that don't need clean up.
+	if tr.DeletionTimestamp != nil || !taskRunCleanup(tr) {
+		return false, nil
+	}
+
+	now := tc.clock.Now()
+	t, err := trTimeLeft(tr, &now)
+	if err != nil {
+		return false, err
+	}
+
+	// TTL has expired
+	if *t <= 0 {
+		return true, nil
+	}
+
+	tc.TrEnqueueAfter(tr, *t)
+	return false, nil
+}
+
+func IsTaskRunSucceeded(tr *apispipeline.TaskRun) bool {
+	for _, con := range tr.Status.Conditions {
+		if con.Type == apis.ConditionSucceeded && con.Status == v1.ConditionTrue {
+			return true
+		}
+	}
+	return false
+}
+
+func getFinishAndExpireTime(tr *apispipeline.TaskRun) (*time.Time, *time.Time, error) {
+	if !taskRunCleanup(tr) {
+		return nil, nil, fmt.Errorf("taskRun %s/%s should not be cleaned up", tr.Namespace, tr.Name)
+	}
+	finishAt, err := taskRunFinishTime(tr)
+	if err != nil {
+		return nil, nil, err
+	}
+	finishAtUTC := finishAt.UTC()
+	expireAtUTC := finishAtUTC.Add(tr.Spec.ExpirationSecondsTTL.Duration * time.Second)
+	return &finishAtUTC, &expireAtUTC, nil
+}
+
+func trTimeLeft(tr *apispipeline.TaskRun, since *time.Time) (*time.Duration, error) {
```

Comment:
I like how this is broken up into lots of small functions - one thing that we'd definitely want to see in this PR is as much unit test coverage as we can get - the easiest way to do this (which usually leads to the best code!) is to create unit tests for individual functions like these

Reply:
OK, It's necessary!